types/webmidi/index.d.ts

```diff
@@ -4,193 +4,234 @@
-     * queue. The implementation will need to ensure the MIDI stream is left in a good
-     * state, so if the output port is in the middle of a sysex message, a sysex
-     * termination byte (0xf7) should be sent.
-     */
-    clear(): void;
-  }
-
-  interface MIDIMessageEvent extends Event {
-    /**
-     * A timestamp specifying when the event occurred.
-     */
-    receivedTime: number;
-
-    /**
-     * A Uint8Array containing the MIDI data bytes of a single MIDI message.
-     */
-    data: Uint8Array;
-  }
-
-  interface MIDIMessageEventInit extends EventInit {
-    /**
-     * A timestamp specifying when the event occurred.
-     */
-    receivedTime: number;
-
-    /**
-     * A Uint8Array containing the MIDI data bytes of a single MIDI message.
-     */
-    data: Uint8Array;
-  }
-
-  interface MIDIConnectionEvent extends Event {
-    /**
-     * The port that has been connected or disconnected.
-     */
-    port: MIDIPort;
-  }
+    requestMIDIAccess(options?: WebMidi.MIDIOptions): Promise<WebMidi.MIDIAccess>;
+}
 
-  interface MIDIConnectionEventInit extends EventInit {
-    /**
-     * The port that has been connected or disconnected.
-     */
-    port: MIDIPort;
-  }
+declare namespace WebMidi {
+    interface MIDIOptions {
+        /**
+         * This member informs the system whether the ability to send and receive system
+         * exclusive messages is requested or allowed on a given MIDIAccess object.
+         */
+        sysex: boolean;
+    }
+
+    /**
+     * This is a maplike interface whose value is a MIDIInput instance and key is its
+     * ID.
+     */
+    type MIDIInputMap = Map<string, MIDIInput>;
+
+    /**
+     * This is a maplike interface whose value is a MIDIOutput instance and key is its
+     * ID.
+     */
+    type MIDIOutputMap = Map<string, MIDIOutput>;
+
+    interface MIDIAccess extends EventTarget {
+        /**
+         * The MIDI input ports available to the system.
+         */
+        inputs: MIDIInputMap;
+
+        /**
+         * The MIDI output ports available to the system.
+         */
+        outputs: MIDIOutputMap;
+
+        /**
+         * The handler called when a new port is connected or an existing port changes the
+         * state attribute.
+         */
+        onstatechange(e: MIDIConnectionEvent): void;
+
+        addEventListener(
+            type: 'statechange',
+            listener: (this: this, e: MIDIConnectionEvent) => any,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+        addEventListener(
+            type: string,
+            listener: EventListenerOrEventListenerObject,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+
+        /**
+         * This attribute informs the user whether system exclusive support is enabled on
+         * this MIDIAccess.
+         */
+        sysexEnabled: boolean;
+    }
+
+    type MIDIPortType = 'input' | 'output';
```

Comment:
Is this still necessary since it's now directly defined on the implementors of `MIDIPort`?

Reply:
See below.